### PR TITLE
Feat : resource namespace change

### PR DESCRIPTION
**Pull request by Mryashbhardwaj**

### CODE REVIEW

File: client/cmd/resource/change_namespace.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	_, err = resourceRunServiceClient.ChangeResourceNamespace(conn.GetContext(), request)
+	return err
+}
+
+func (c *changeNamespaceCommand) PostRunE(_ *cobra.Command, args []string) error {
+	c.logger.Info("\n[INFO] Moving resource in filesystem")
+	resourceName := args[0]
+	fs := afero.NewOsFs()
+	var source, destination string
+	for _, namespace := range c.clientConfig.Namespaces {
```
Comment (deryrahman) [inline client/cmd/resource/change_namespace.go:114]: Do we allow to move the resource to the existing namespace? If it's allowed, then we couldn't just rename the folder. But if it's not allowed, then how do we ensure the namespace destination is newly created and don't have any preoccupied resource spec?

File: core/resource/handler/v1beta1/resource.go [hunk 2]
```diff
@@ -242,6 +243,34 @@ func (rh ResourceHandler) UpdateResource(ctx context.Context, req *pb.UpdateReso
...
+	if err != nil {
+		return nil, errors.GRPCErr(err, "failed to adapt to new tenant")
+	}
+
+	store, err := resource.FromStringToStore(req.GetDatastoreName())
+	if err != nil {
+		return nil, errors.GRPCErr(err, "invalid Datastore Name")
+	}
+
+	response, err := rh.service.Get(ctx, tnnt, store, req.GetResourceName())
```
Comment (irainia) [inline core/resource/handler/v1beta1/resource.go:262]: Please put this as part of the `service.ChangeNamespace`. Try to minimize business flow as much as possible in this handler.

File: core/resource/resource.go [hunk 3]
```diff
@@ -94,6 +94,10 @@ func (r *Resource) UpdateURN(urn string) error {
 	return errors.InvalidArgument(EntityResource, "urn already present for "+r.FullName())
 }
 
+func (r *Resource) UpdateTenant(tnnt tenant.Tenant) {
+	r.tenant = tnnt
```
Comment (irainia) [inline core/resource/resource.go:98]: Please add some validation in here. An example, whether we will allow update in project or we just restrict to namespace only.

File: core/scheduler/service/replay_service.go [hunk 4]
```diff
@@ -68,7 +68,12 @@ func (r ReplayService) GetReplayList(ctx context.Context, projectName tenant.Pro
 }
 
 func (r ReplayService) GetReplayByID(ctx context.Context, replayID uuid.UUID) (*scheduler.ReplayWithRun, error) {
-	return r.replayRepo.GetReplayByID(ctx, replayID)
+	replayWithRun, err := r.replayRepo.GetReplayByID(ctx, replayID)
+	if err != nil {
+		return nil, err
+	}
+	replayWithRun.Runs = scheduler.JobRunStatusList(replayWithRun.Runs).GetSortedRunsByScheduledAt()
```
Comment (irainia) [inline core/scheduler/service/replay_service.go:75]: do we need this changes for moving namespace for a resource?

File: client/cmd/resource/change_namespace.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		ResourceName:     resourceName,
+		NewNamespaceName: c.newNamespace,
+	}
+
+	_, err = resourceRunServiceClient.ChangeResourceNamespace(conn.GetContext(), request)
+	return err
+}
+
+func (c *changeNamespaceCommand) PostRunE(_ *cobra.Command, args []string) error {
+	c.logger.Info("\n[INFO] Moving resource in filesystem")
```
Comment (irainia) [inline client/cmd/resource/change_namespace.go:110]: If the main idea is to move namespace in local as well, what do you think we do some validation first on the local state before sending request to the server? For example, does the "old" namespace exist in local?

However, if the main idea is "moving namespace in local is *optional*", then we don't need to do that validation before sending to server. But, we still need to check whether both exist in local before renaming things.

File: client/cmd/resource/change_namespace.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	resourceName := args[0]
+	fs := afero.NewOsFs()
+	var source, destination string
+	for _, namespace := range c.clientConfig.Namespaces {
+		if namespace.Name == c.oldNamespace {
+			for _, datastore := range namespace.Datastore {
+				if datastore.Type == c.dataStore {
+					source = "./" + datastore.Path + "/" + resourceName
```
Comment (irainia) [inline client/cmd/resource/change_namespace.go:118]: I don't think we need to add `./` in this line (and below line for destination). Also, for path, please use `path.Join()` instead.

**Review by irainia — CHANGES_REQUESTED** [review]

File: client/cmd/resource/change_namespace.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+					destination = "./" + datastore.Path + "/" + resourceName
+					break
+				}
+			}
+		}
+	}
+
+	c.logger.Info(fmt.Sprintf("\t* Old Path : '%s' \n\t* New Path : '%s' \n", source, destination))
+
+	err := fs.Rename(source, destination)
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go]: we will be logging this as an error , but the job will be moved in the optimus server

File: client/cmd/resource/change_namespace.go (hunk 1 above)
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go]: we will always be creating a new directory in the new namespace with the same directory structure if not already exists, 

But if the job folder also exists with the same name then we give error 

`Error: rename file exists`

File: internal/store/postgres/resource/repository.go [hunk 8]
```diff
@@ -52,6 +52,22 @@ func (r Repository) Update(ctx context.Context, resourceModel *resource.Resource
 	return nil
 }
 
+func (r Repository) ChangeNamespace(ctx context.Context, res *resource.Resource, newTenant tenant.Tenant) error {
+	updateResource := `UPDATE resource SET namespace_name=$1, updated_at=now()
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply internal/store/postgres/resource/repository.go:56]: resources namespace is an optimus construct, changing a resource namespace should not have impact on a jobs access to that resource to that on BQ

File: client/cmd/resource/change_namespace.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+// NewChangeNamespaceCommand initializes resource namespace change command
+func NewChangeNamespaceCommand() *cobra.Command {
+	l := logger.NewClientLogger()
+	changeNamespace := &changeNamespaceCommand{
+		logger: l,
+	}
+	cmd := &cobra.Command{
+		Use:      "change-namespace",
+		Short:    "Create namespace of a resource",
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go:49]: ohh i missed that, fixing it

File: core/scheduler/service/replay_service.go (hunk 4 above)
Comment (Mryashbhardwaj) [reply core/scheduler/service/replay_service.go]: no , i think this has come from base branch , will fix this

File: core/resource/resource.go (hunk 3 above)
Comment (Mryashbhardwaj) [reply core/resource/resource.go:98]: for now, this is supposed to change the complete tenant

File: client/cmd/resource/change_namespace.go (hunk 5 above)
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go:118]: we see moving in local as an added feature , thats why we are doing it after the main purpose i.e. moving the resource on the server

File: client/cmd/resource/change_namespace.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+		}
+		c.logger.Warn("[info] removed resource spec from current namespace directory")
+		c.logger.Warn("[info] run `optimus resource export` on the new namespace repo, to fetch the newly moved resource.")
+		c.logger.Info("[OK] Resource moved successfully")
+		return nil
+	}
+
+	var relativeResourcePath string
+	splitComp := strings.Split(resourceSpec.Path, oldNamespaceConfig.Path)
+	if !(len(splitComp) > 1) {
```
Comment (deryrahman) [inline client/cmd/resource/change_namespace.go:155]: nit: `len(splitComp) <= 1`

File: client/cmd/resource/change_namespace.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+		return errors.NewError(errors.ErrInternalError, "change-namespace", "unable to parse resource spec path")
+	}
+	relativeResourcePath = splitComp[1]
+
+	c.logger.Info(fmt.Sprintf("\t* Old Path : '%s' \n\t* New Path : '%s' \n", resourceSpec.Path, newNamespaceConfig.Path+relativeResourcePath))
```
Comment (deryrahman) [inline client/cmd/resource/change_namespace.go:160]: nit: store this `newNamespaceConfig.Path+relativeResourcePath` in variable

**Review by deryrahman — CHANGES_REQUESTED** [review]

File: core/resource/handler/v1beta1/resource.go (hunk 2 above)
Comment (Mryashbhardwaj) [reply core/resource/handler/v1beta1/resource.go]: sure

File: client/cmd/resource/change_namespace.go (hunk 10 above)
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go]: done

File: core/resource/resource.go (hunk 3 above)
Comment (irainia) [reply core/resource/resource.go:98]: sure

File: client/cmd/resource/change_namespace.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+	if err != nil {
+		return errors.Wrap(tenant.EntityNamespace, "unable to find resource in old namespace", err)
+	}
+
+	fs := afero.NewOsFs()
+	newNamespaceConfig, err := c.getResourceDatastoreConfig(c.newNamespace, c.dataStore)
+	if err != nil || newNamespaceConfig.Path == "" {
+		c.logger.Warn("[warn] new namespace not recognised for Resources")
+		c.logger.Warn("[info] removing resource from old namespace")
+		err = fs.RemoveAll(resourceSpec.Path)
```
Comment (deryrahman) [inline client/cmd/resource/change_namespace.go:142]: are we removing the resource spec in filesystem although the new namespace path is not exist?

File: client/cmd/resource/change_namespace.go [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+			return errors.NewError(errors.ErrInternalError, "change-namespace", "unable to remove resource from old namespace")
+		}
+		c.logger.Warn("[info] removed resource spec from current namespace directory")
+		c.logger.Warn("[info] run `optimus resource export` on the new namespace repo, to fetch the newly moved resource.")
+		c.logger.Info("[OK] Resource moved successfully")
+		return nil
+	}
+
+	var relativeResourcePath string
+	splitComp := strings.Split(resourceSpec.Path, oldNamespaceConfig.Path)
```
Comment (deryrahman) [inline client/cmd/resource/change_namespace.go:154]: `newResourcePath = strings.Replace(resourceSpec.Path, oldNamespaceConfig.Path, newNamespaceConfig.Path, 1)`, we can use this and remove line 154-159

File: client/cmd/resource/change_namespace.go (hunk 12 above)
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go:142]: yes

File: client/cmd/resource/change_namespace.go (hunk 13 above)
Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go:154]: nice idea

Comment (Mryashbhardwaj) [reply client/cmd/resource/change_namespace.go]: done

**Review by deryrahman — APPROVED** [review]

**Review by irainia — APPROVED** [review]

**Pull request merged**